pkg/bootstrap/platform/gcp.go

```diff
@@ -43,48 +48,64 @@ var (
 )
 
 var (
-	clusterNameFn = func() (string, error) {
-		cn, err := metadata.InstanceAttributeValue("cluster-name")
-		if err != nil {
-			return "", err
-		}
-		return cn, nil
+	clusterNameFn = func(e *gcpEnv) (string, error) {
+		f := func() (string, error) { return metadata.InstanceAttributeValue("cluster-name") }
+		return getCacheMetadata(&e.clusterName, f)
 	}
-	clusterLocationFn = func() (string, error) {
-		cl, err := metadata.InstanceAttributeValue("cluster-location")
-		if err == nil {
-			return cl, nil
+	clusterLocationFn = func(e *gcpEnv) (string, error) {
+		f := func() (string, error) { return metadata.InstanceAttributeValue("cluster-location") }
+		if location, err := getCacheMetadata(&e.location, f); err == nil {
+			return location, nil
 		}
-		return metadata.Zone()
+		return getCacheMetadata(&e.location, metadata.Zone)
 	}
-	instanceTemplateFn = func() (string, error) {
-		it, err := metadata.InstanceAttributeValue("instance-template")
-		if err != nil {
-			return "", err
-		}
-		return it, nil
+	instanceNameFn = func(e *gcpEnv) (string, error) {
+		return getCacheMetadata(&e.instanceName, metadata.InstanceName)
 	}
-	createdByFn = func() (string, error) {
-		cb, err := metadata.InstanceAttributeValue("created-by")
-		if err != nil {
-			return "", err
-		}
-		return cb, nil
+	instanceTemplateFn = func(e *gcpEnv) (string, error) {
+		f := func() (string, error) { return metadata.InstanceAttributeValue("instance-template") }
+		return getCacheMetadata(&e.instanceTemplate, f)
+	}
+	createdByFn = func(e *gcpEnv) (string, error) {
+		f := func() (string, error) { return metadata.InstanceAttributeValue("created-by") }
+		return getCacheMetadata(&e.instanceTemplate, f)
+	}
+	shouldFillMetadata = metadata.OnGCE
+	projectIDFn        = func(e *gcpEnv) (string, error) {
+		return getCacheMetadata(&e.projectID, metadata.ProjectID)
+	}
+	numericProjectIDFn = func(e *gcpEnv) (string, error) {
+		return getCacheMetadata(&e.numericProjectID, metadata.NumericProjectID)
+	}
+	instanceIDFn = func(e *gcpEnv) (string, error) {
+		return getCacheMetadata(&e.instanceID, metadata.InstanceID)
 	}
 )
 
 type shouldFillFn func() bool
-type metadataFn func() (string, error)
+type metadataFn func(e *gcpEnv) (string, error)
 
 type gcpEnv struct {
-	shouldFillMetadata shouldFillFn
-	projectIDFn        metadataFn
-	numericProjectIDFn metadataFn
-	locationFn         metadataFn
-	clusterNameFn      metadataFn
-	instanceIDFn       metadataFn
-	instanceTemplateFn metadataFn
-	createdByFn        metadataFn
+	projectID        string
+	numericProjectID string
+	location         string
+	clusterName      string
+	instanceName     string
+	instanceID       string
+	instanceTemplate string
+	createdBy        string
+}
+
+// Helper for compute metadata getters. Caches and returns f() if valid.
+func getCacheMetadata(s *string, f func() (string, error)) (string, error) {
```

Comment:
I wrote it this way due to the mentioned potential flakiness of metadata. With the above, the same problem as using `once` would still be there as there'd be no way of addressing an initial failed call, so I thought caching would a better approach since it would try to repull until success. If we can ignore initial failures (e.g. empty attributes) then the above one-pass would be a better choice. 
I agree that the cache is a bit clunky, just didn't wanna write the logic into ever metadata retrieval func.

Reply:
Is the same gcpEnv ever used concurrently? I don't see existing code that would require Metadata() to be blocking

Reply:
I don't believe there is any concurrent usage of `gcpEnv`; on the other hand `gcpEnv` is stateless (if we ignore the package-level state added for the `GCP_METADATA` bit). If we are going to add state, it'd be nice not to have to worry about potential future code changes that might introduce concurrency issues.

If you are concerned about blocking/caching issues, you could also try calling `metadataFns` you need for `Labels()` using retry to avoid flakes as well and only focus on that section of the code.